src/shadowsocks_config.ts

```diff
@@ -338,3 +338,63 @@ export const SIP002_URI = {
     return `ss://${userInfo}@${uriHost}:${port.data}/${queryString}${hash}`;
   },
 };
+
+export interface OnlineConfig {
+  // URL endpoint to retrieve a Shadowsocks configuration.
+  readonly url: string;
+  // Any additional configuration (e.g. `certFp`, `httpMethod`, etc.) may be stored here.
+  readonly extra: {[key: string]: string};
+}
+
+// Ref: https://github.com/shadowsocks/shadowsocks-org/issues/89
+export const SIP008_URI = {
+  PROTOCOL: 'ssconf',
+
+  validateProtocol: (uri: string) => {
+    if (!uri || !uri.startsWith(SIP008_URI.PROTOCOL)) {
+      throw new InvalidUri(`URI must start with "${SIP008_URI.PROTOCOL}"`);
+    }
+  },
+
+  parse: (uri: string): OnlineConfig => {
+    SIP008_URI.validateProtocol(uri);
+
+    // URL parser for expedience, replacing the protocol "ssconf" with "https" to ensure correct
+    // results, otherwise browsers like Safari fail to parse it.
+    const inputForUrlParser =
+        `https${decodeURIComponent(uri.substring(SIP008_URI.PROTOCOL.length))}`;
+    // The built-in URL parser throws as desired when given URIs with invalid syntax.
+    const urlParserResult = new URL(inputForUrlParser);
+    // Use ValidatedConfigFields subclasses (Host, Port, Tag) to throw on validation failure.
+    const uriFormattedHost = urlParserResult.hostname;
```

Comment:
If we do a regex replace like `...hostname.replace(/\[(.*)\]/, '$1')` here then we don't need the try-catch

Reply:
unfortunately that doesn't quite work because some implementations fail without the brackets. I don't remember for which browsers/webviews we added the try catch in SIP002, but I thought it would make sense to replicate it here.